capnp 0.9.0 failed to compile on Majove

👋  just found the latest release 0.9.0 does not compile on Mojave (macOS 10.14), error log is below:
<details>
<summary>build error</summary>

```
In file included from /tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/mutex.h:28:
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:59:64: error: use of undeclared identifier '__builtin_FILE'
  constexpr SourceLocation(Badge = Badge{}, const char* file = __builtin_FILE(),
                                                               ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:60:26: error: use of undeclared identifier '__builtin_FUNCTION'
      const char* func = __builtin_FUNCTION(), uint line = __builtin_LINE(),
                         ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:60:60: error: use of undeclared identifier '__builtin_LINE'
      const char* func = __builtin_FUNCTION(), uint line = __builtin_LINE(),
                                                           ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:61:21: error: use of undeclared identifier '__builtin_COLUMN'
      uint column = KJ_CALLER_COLUMN())
                    ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:27:28: note: expanded from macro 'KJ_CALLER_COLUMN'
#define KJ_CALLER_COLUMN() __builtin_COLUMN()
                           ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:59:13: error: constexpr constructor must initialize all members
  constexpr SourceLocation(Badge = Badge{}, const char* file = __builtin_FILE(),
            ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:77:15: note: member not initialized by constructor
  const char* fileName;
              ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:78:15: note: member not initialized by constructor
  const char* function;
              ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:79:8: note: member not initialized by constructor
  uint lineNumber;
       ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:80:8: note: member not initialized by constructor
  uint columnNumber;
       ^
5 errors generated.
In file included from /tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.c++:22:
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:59:64: error: use of undeclared identifier '__builtin_FILE'
  constexpr SourceLocation(Badge = Badge{}, const char* file = __builtin_FILE(),
                                                               ^
make[2]: *** [src/kj/CMakeFiles/kj.dir/mutex.c++.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:60:26: error: use of undeclared identifier '__builtin_FUNCTION'
      const char* func = __builtin_FUNCTION(), uint line = __builtin_LINE(),
                         ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:60:60: error: use of undeclared identifier '__builtin_LINE'
      const char* func = __builtin_FUNCTION(), uint line = __builtin_LINE(),
                                                           ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:61:21: error: use of undeclared identifier '__builtin_COLUMN'
      uint column = KJ_CALLER_COLUMN())
                    ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:27:28: note: expanded from macro 'KJ_CALLER_COLUMN'
#define KJ_CALLER_COLUMN() __builtin_COLUMN()
                           ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:59:13: error: constexpr constructor must initialize all members
  constexpr SourceLocation(Badge = Badge{}, const char* file = __builtin_FILE(),
            ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:77:15: note: member not initialized by constructor
  const char* fileName;
              ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:78:15: note: member not initialized by constructor
  const char* function;
              ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:79:8: note: member not initialized by constructor
  uint lineNumber;
       ^
/tmp/capnp-20210815-80151-159a7vv/capnproto-c++-0.9.0/src/kj/source-location.h:80:8: note: member not initialized by constructor
  uint columnNumber;
       ^
5 errors generated.
make[2]: *** [src/kj/CMakeFiles/kj.dir/source-location.c++.o] Error 1
```

</details>

The cmake command 
```
cmake .. -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/capnp/0.9.0 -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DBUILD_TESTING=OFF -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
```

relates to https://github.com/Homebrew/homebrew-core/pull/83289

--- a/c++/src/kj/source-location.h
+++ b/c++/src/kj/source-location.h
@@ -23,20 +23,33 @@
 
 #include "string.h"
 
-#if __clang__ && __clang_major__ >= 9
-#define KJ_CALLER_COLUMN() __builtin_COLUMN()
 // GCC does not implement __builtin_COLUMN() as that's non-standard but MSVC & clang do.
 // MSVC does as of version https://github.com/microsoft/STL/issues/54) but there's currently not any
 // pressing need for this for MSVC & writing the write compiler version check is annoying.
+// Checking for clang version is problematic due to the way that XCode lies about __clang_major__.
+// Instead we use __has_builtin as the feature check to check clang.
+// Context: https://github.com/capnproto/capnproto/issues/1305
+#ifdef __has_builtin
+#if __has_builtin(__builtin_COLUMN)
+#define KJ_CALLER_COLUMN() __builtin_COLUMN()
+#else
+#define KJ_CALLER_COLUMN() 0
+#endif
 #else
 #define KJ_CALLER_COLUMN() 0
 #endif
 
 #if __cplusplus > 201703L
 #define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
-#elif __clang__ && __clang_major__ >= 9
+#elif defined(__has_builtin)
 // Clang 9 added these builtins: https://releases.llvm.org/9.0.0/tools/clang/docs/LanguageExtensions.html
-#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
+// Use __has_builtin as the way to detect this because __clang_major__ is unreliable (see above
+// about issue with Xcode-provided clang).
+#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION (   \
+  __has_builtin(__builtin_FILE) &&               \
+  __has_builtin(__builtin_LINE) &&               \
+  __has_builtin(__builtin_FUNCTION)              \
+  )
 #elif __GNUC__ >= 5
 // GCC 5 supports the required builtins: https://gcc.gnu.org/onlinedocs/gcc-5.1.0/gcc/Other-Builtins.html
 #define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1

